### PR TITLE
Investigate Kubernetes pod CPU usage

### PR DESCRIPTION
**Pull request by JamesGardiner**
### What is the context of this PR?
Describe what you have changed and why, link to other PRs or Issues as appropriate.

### How to review 
Describe the steps required to test the changes (include screenshots if appropriate).

### CODE REVIEW

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Fluentd Performance
+
+As part of the testing, fluentd pods were observed to be using around 450 mCPU when under sustained load. Their requested cpu time is only 100 mCPU. This means that, when under load, fluentd log forwarding is using roughly half a core of available CPU. Metrics gathered during recent stress tests show the same level of resource usage. Unfortunately, it was not feasible to get a correct comparison of the impact that disabling app logging has at the cluster level, as the tests that were run when fluentd was disabled, all had multiple `kube-system` pods placed on Runner nodes, however, given the impact that we see when 648 mCPU of additional CPU is requested by `kube-system` pods, it is likely that the increased usage by fluentd will have a negative impact. It may be useful to look at ways to tune fluentd to increase its performance.
```
Comment (ajmaddaford) [inline doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:29]: This is a very long sentence!

> Unfortunately, it was not feasible to get a correct comparison of the impact that disabling app logging has at the cluster level, as the tests that were run when fluentd was disabled, all had multiple `kube-system` pods placed on Runner nodes, however, given the impact that we see when 648 mCPU of additional CPU is requested by `kube-system` pods, it is likely that the increased usage by fluentd will have a negative impact.

**Review by ajmaddaford — COMMENTED** [review]
The commit message should be changed when this is merged, as it's not relevant which number investigation it is (they're only numbered to order them).

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md (hunk 1 above)
Comment (JamesGardiner) [reply doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:29]: I'll reword it!

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+# Summary
+
+## CPU Allocation
+
+In the k8s cluster, the allocation of CPU cores is as follows:
+
+- Each node has 3,920 mCPU allocatable.
+- On each node, there is _always_ a minimum of 203 mCPU requested by pods in the `kube-system` namespace, leaving 3,717 mCPU for other pods.
+- 648 mCPU of cores are requested by kube-system pods once per cluster. These can be put on a single node, or split across nodes if there are available CPU cores.
```
[the remark this answers is not in the document]
Comment (JamesGardiner) [reply doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:13]: from my testing, and from looking at the stress test data it appears that yes, this is a static request per cluster. Additionally, I did  not see any of the `kube-system` pods use above their requested resource count apart from `fluentd`.

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+A number of benchmark tests were carried out to investigate the CPU usage of kubernetes pods, and how this effects Questionnaire Runner.
+
+# Summary
+
+## CPU Allocation
+
+In the k8s cluster, the allocation of CPU cores is as follows:
+
+- Each node has 3,920 mCPU allocatable.
```
Comment (MebinAbraham) [inline doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:11]: I think it's worth mentioning this is when the provisioned machine has 4 cores.

**Review by MebinAbraham — COMMENTED** [review]
Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md (hunk 1 above)
Comment (JamesGardiner) [reply doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md]: Done

**Comment by JamesGardiner** [comment]
> Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?

I did consider not committing them as they aren't actually that relevant to this investigation. I decided to include them in the repo as they are still valid outputs, so could potentially be of use at some point. Happy to remove them if we feel that's more appropriate.

**Comment by MebinAbraham** [comment]
> > Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?
> 
> I did consider not committing them as they aren't actually that relevant to this investigation. I decided to include them in the repo as they are still valid outputs, so could potentially be of use at some point. Happy to remove them if we feel that's more appropriate.

I think because there is no context in the outputs apart from runner version, I don't think it gives us much to go on even if we were to look at it in the future. i.e I can't tell what CPU config was used for these tests. Are they worth putting in the doc?. What were the actual tests that generated these outputs?.

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md (hunk 3 above)
Comment (JamesGardiner) [reply doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:11]: done

**Comment by JamesGardiner** [comment]
> > > Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?
> > 
> > 
> > I did consider not committing them as they aren't actually that relevant to this investigation. I decided to include them in the repo as they are still valid outputs, so could potentially be of use at some point. Happy to remove them if we feel that's more appropriate.
> 
> I think because there is no context in the outputs apart from runner version, I don't think it gives us much to go on even if we were to look at it in the future. i.e I can't tell what CPU config was used for these tests. Are they worth putting in the doc?. What were the actual tests that generated these outputs?.

They are output from running the benchmark tests against the runner instances.

**Comment by MebinAbraham** [comment]
> > > > Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?
> > > 
> > > 
> > > I did consider not committing them as they aren't actually that relevant to this investigation. I decided to include them in the repo as they are still valid outputs, so could potentially be of use at some point. Happy to remove them if we feel that's more appropriate.
> > 
> > 
> > I think because there is no context in the outputs apart from runner version, I don't think it gives us much to go on even if we were to look at it in the future. i.e I can't tell what CPU config was used for these tests. Are they worth putting in the doc?. What were the actual tests that generated these outputs?.
> 
> They are output from running the benchmark tests against the runner instances.

But what was the baseline vs investigation variables?

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajmaddaford) [reply doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:13]: I think it would be worth this information to the doc.

**Comment by JamesGardiner** [comment]
> > > > > Unless the benchmark stats are represented in the document, I don't think they should be committed. Thoughts?
> > > > 
> > > > 
> > > > I did consider not committing them as they aren't actually that relevant to this investigation. I decided to include them in the repo as they are still valid outputs, so could potentially be of use at some point. Happy to remove them if we feel that's more appropriate.
> > > 
> > > 
> > > I think because there is no context in the outputs apart from runner version, I don't think it gives us much to go on even if we were to look at it in the future. i.e I can't tell what CPU config was used for these tests. Are they worth putting in the doc?. What were the actual tests that generated these outputs?.
> > 
> > 
> > They are output from running the benchmark tests against the runner instances.
> 
> But what was the baseline vs investigation variables?

As agreed offline, I will remove these from the PR

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Fluentd Performance
+
+Fluentd pods were observed to be using around 450 mCPU when under sustained load. Their requested cpu time is only 100 mCPU. Metrics gathered during recent stress tests show similar levels of CPU resource usage. Unfortunately, it was not feasible to measure the impact disabling app logging has at the cluster level; all of the tests run with fluentd disabled had multiple `kube-system` pods on Runner nodes. It is likely that that the increase in CPU resource used by fluentd under high load has a negative impact on response times. It may be beneficial to look at fluentd performance tuning.
+
+## Decision
+
+1. Do not increase the runner CPU request.
+1. Test environments should use 3 runner pods on 3 nodes rather than the 2 pods that are currently deployed.
+1. Investigate the feasibility of tuning fluentd.
```
Comment (ajmaddaford) [inline doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:35]: Should one of the decisions be to test different machine types?

**Review by MebinAbraham — APPROVED** [review]

File: doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+# Summary
+
+## CPU Allocation
+
+Our nodes (which are 4 core virtual machines) allocate CPU cores as follows:
+
+- Each node has 3,920 mCPU allocatable.
+- On each node, there is _always_ a minimum of 203 mCPU requested by pods in the `kube-system` namespace, leaving 3,717 mCPU for other pods.
+- 648 mCPU of cores are requested by kube-system pods once per cluster. These can be put on a single node, or split across nodes if there are available CPU cores.
+- From the testing carried out, it appears that the 648mCPU request is static per Kubernetes cluster.
```
Comment (ajmaddaford) [inline doc/performance-investigations/0011-increase-pod-cpu-allocation/summary.md:14]: This should also mention the stress test. 
```suggestion
- From the testing carried out and previous stress tests, it appears that the 648 mCPU request is static per Kubernetes cluster.
```

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**